File: py/src/fit/session.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .advisor import Advisor
from .reward import RewardScorer
from .types import Advice, Reward, Trace


@dataclass
class SessionConfig:
    mode: str = "one-shot"
    max_steps: int = 10
    reward_threshold: float = 1.0


class Session:
    """Orchestrates advisor → frontier → reward → trace cycle."""

    def __init__(
        self,
        advisor: Advisor,
        adapter: Any,
        scorer: RewardScorer,
        config: SessionConfig | None = None,
    ) -> None:
        self._advisor = advisor
        self._adapter = adapter
        self._scorer = scorer
        self._config = config or SessionConfig()

    def run(self, prompt: str, context: dict[str, Any] | None = None) -> tuple[str, Reward, Trace]:
        """Run a one-shot session. Returns (output, reward, trace)."""
        import uuid
        from datetime import datetime, timezone

        ctx = context or {}
        session_id = str(uuid.uuid4())

        # Advise
        try:
            advice = self._advisor.generate_advice({"prompt": prompt, **ctx})
        except Exception:
            advice = Advice(domain="unknown", steering_text="", confidence=0.0)

        # Frontier
        output, frontier_meta = self._adapter.call(prompt, advice)

        # Score
        try:
            reward = self._scorer.score(output, ctx)
        except Exception:
            reward = Reward(score=float("nan"), breakdown={})

        trace = Trace(
            id=str(uuid.uuid4()),
            session_id=session_id,
            timestamp=datetime.now(timezone.utc).isoformat(),
            input={"prompt": prompt, "context": ctx},
            advice=advice,
            frontier=frontier_meta,
            reward=reward,
        )

        return output, reward, trace

File: php/src/Session.php
<?php

declare(strict_types=1);

namespace Hop\Fit;

class Session
{
    public function __construct(
        private AdvisorInterface $advisor,
        private AdapterInterface $adapter,
        private RewardScorerInterface $scorer,
    ) {}

    /**
     * @return array{string, Reward, Trace}
     */
    public function run(string $prompt, array $context = []): array
    {
        $sessionId = bin2hex(random_bytes(16));

        try {
            $advice = $this->advisor->generateAdvice(['prompt' => $prompt] + $context);
        } catch (\Throwable) {
            $advice = new Advice('unknown', '', 0.0);
        }

        [$output, $frontierMeta] = $this->adapter->call($prompt, $advice);

        try {
            $reward = $this->scorer->score($output, $context);
        } catch (\Throwable) {
            $reward = new Reward(NAN, []);
        }

        $trace = new Trace(
            id: bin2hex(random_bytes(16)),
            sessionId: $sessionId,
            timestamp: gmdate('c'),
            input: ['prompt' => $prompt] + $context,
            advice: $advice,
            frontier: $frontierMeta,
            reward: $reward,
        );

        return [$output, $reward, $trace];
    }
}
